lib/ace/ext/papeeria/katex-previewer.coffee

```diff
@@ -219,113 +233,175 @@ define((require, exports, module) ->
-        else
-          curSequenceIndex = EquationRangeHandler.BEGIN_EQUATION_TOKEN_SEQUENCE.length - 1
-          curEquationStart = null
-      return curEquationStart
 
-    getEquationEnd: (tokenIterator) ->
-      # if tokenIterator is initially on the empty line, its current token is null
-      if not tokenIterator.getCurrentToken()?
-        tokenIterator.stepBackward()
-        # if tokenIterator is still null, then we're at the end of a file
-        if not tokenIterator.getCurrentToken()?
-          return null
-      else
-        # following cycle pushes tokenIterator to the start of
-        # ending sequence, if it is inside one
-        # The loop isn't executed, if current token is null, because:
-        #   a) we don't need to -- if `tokenIterator` is initially on
-        #      the empty line, then it is guaranteed not to be inside start sequence
-        #   b) it can cause bugs -- if `tokenIterator` is initially after the end of a document,
-        #      current token is null, and after stepping backward `tokenIterator` is on the
-        #      end of a document. If end sequence happens to be there, then equation end is
-        #      then found without any problem, whereas in this case null should be returned
-        for token in EquationRangeHandler.END_EQUATION_TOKEN_SEQUENCE.slice(0).reverse()
-          if EquationRangeHandler.equalTokens(token, tokenIterator.getCurrentToken())
-            tokenIterator.stepBackward()
-
-      curSequenceIndex = 0
-      curEquationStart = null
-      while curSequenceIndex < EquationRangeHandler.END_EQUATION_TOKEN_SEQUENCE.length
-        curToken = tokenIterator.stepForward()
-        if not curToken
-          return null
-        if EquationRangeHandler.equalTokens(
-            EquationRangeHandler.END_EQUATION_TOKEN_SEQUENCE[curSequenceIndex],
-            curToken)
-          if curSequenceIndex == 0
-            curEquationStart = tokenIterator.getCurrentTokenPosition()
-          curSequenceIndex += 1
-        else
-          curSequenceIndex = 0
-          curEquationStart = null
-      return curEquationStart
+        tokenSequenceFinder.progressIndices(curToken)
+
+        maybeFinishedSequenceId = tokenSequenceFinder.getMaybeFinishedSequenceId()
+        if maybeFinishedSequenceId?
+          { row: curTokenRow, column: curTokenColumn } = tokenIterator.getCurrentTokenPosition()
+          curTokenLength = curToken.value.length
+          finishedSequence = boundarySequences[maybeFinishedSequenceId]
+          finishedSequenceStringLength = (token.value for token in finishedSequence).join("").length
+
+          equationOuterBoundary = {
+            row: curTokenRow
+            column: curTokenColumn + (if start then 0 else curTokenLength)
+          }
+          equationInnerBoundary = {
+            row: curTokenRow
+            column: curTokenColumn + (if start then finishedSequenceStringLength else curTokenLength - finishedSequenceStringLength)
```

Comment:
this works because sequences don't allow whitespace, right?  If whitespace was allowed then the sum of the sequence token lengths would not necessary be the length of the sequence